### ISSUE TITLE
"basic" search giving weird results

### ISSUE THREAD

**Issue by Ansa211** [ufal/lindat-kontext#142]
EDIT:
the important question is the one stated in bold in https://github.com/ufal/lindat-kontext/issues/142#issuecomment-348824512 

-----------------------
[Czech Legal Text Treebank, basic search for "vyhláška"](https://lindat.mff.cuni.cz/services/kontext-staging/ansa/view?ctxattrs=word&attr_vmode=visible&pagesize=40&refs=%3Ddoc.id&q=~3qIukf2o&viewmode=kwic&attrs=word%2Clemma%2Ctag%2Cafun%2Cparent%2Cp_form%2Cp_lemma%2Cp_tag%2Cp_afun%2Ceparent%2Cep_form%2Cep_lemma%2Cep_tag%2Cep_afun%2Clc%2Cpos%2Ck%2Cg%2Cc&corpname=legaltext_cs_a&attr_allpos=kw)

Searching for [word form](https://lindat.mff.cuni.cz/services/kontext-staging/ansa/view?ctxattrs=word&attr_vmode=visible&pagesize=40&refs=%3Ddoc.id&q=~v4eHCxb3&viewmode=kwic&attrs=word%2Clemma%2Ctag%2Cafun%2Cparent%2Cp_form%2Cp_lemma%2Cp_tag%2Cp_afun%2Ceparent%2Cep_form%2Cep_lemma%2Cep_tag%2Cep_afun%2Clc%2Cpos%2Ck%2Cg%2Cc&corpname=legaltext_cs_a&attr_allpos=kw) or [lemma](https://lindat.mff.cuni.cz/services/kontext-staging/ansa/view?ctxattrs=word&attr_vmode=visible&pagesize=40&refs=%3Ddoc.id&q=~pWGnT3b4&viewmode=kwic&attrs=word%2Clemma%2Ctag%2Cafun%2Cparent%2Cp_form%2Cp_lemma%2Cp_tag%2Cp_afun%2Ceparent%2Cep_form%2Cep_lemma%2Cep_tag%2Cep_afun%2Clc%2Cpos%2Ck%2Cg%2Cc&corpname=legaltext_cs_a&attr_allpos=kw) ~~or even using a [CQL equivalent of the basic search](https://lindat.mff.cuni.cz/services/kontext-staging/ansa/view?ctxattrs=word&attr_vmode=visible&pagesize=40&refs=%3Ddoc.id&q=~Frx12RNX&viewmode=kwic&attrs=word%2Clemma%2Ctag%2Cafun%2Cparent%2Cp_form%2Cp_lemma%2Cp_tag%2Cp_afun%2Ceparent%2Cep_form%2Cep_lemma%2Cep_tag%2Cep_afun%2Clc%2Cpos%2Ck%2Cg%2Cc&corpname=legaltext_cs_a&attr_allpos=kw) (if `[word="vyhláška" | lc="vyhláška" | lemma="vyhláška" ]` is the correct equivalent)~~ all give correct results, so I am quite baffled.

I tried running the query in anonymous mode to be sure it's not caused by cache. I also restarted kontext-ansa and both of the taskers, but that did not help either.

Currently, I do not know about other treebanks having the same problem. Searching for "část" seems to give correct results; searching for "dům" does not.

**Comment by Ansa211** [comment]
Update:

The CQL equivalent of the basic search is [`[lc="vyhláška"|lemma="(?i)vyhláška"]`](https://lindat.mff.cuni.cz/services/kontext-staging/ansa/view?ctxattrs=word&attr_vmode=visible&pagesize=100&refs=%3Ddoc.id&q=~poxfOs4Q&viewmode=kwic&attrs=word&corpname=legaltext_cs_a&attr_allpos=kw), which gives the same wrong result as the basic search.
Both disjuncts give correct results ([`[lc="vyhláška"]`](https://lindat.mff.cuni.cz/services/kontext-staging/ansa/view?ctxattrs=word&attr_vmode=visible&pagesize=100&refs=%3Ddoc.id&q=~xM3MHLOD&viewmode=kwic&attrs=word&corpname=legaltext_cs_a&attr_allpos=kw), [`[lemma="(?i)vyhláška"]`](https://lindat.mff.cuni.cz/services/kontext-staging/ansa/view?ctxattrs=word&attr_vmode=visible&pagesize=100&refs=%3Ddoc.id&q=~aiIKXoOt&viewmode=kwic&attrs=word&corpname=legaltext_cs_a&attr_allpos=kw)).
So tried to find another disjunction which would also give a wrong result, but I have not been successfull.

**Comment by kosarko** [comment]
isn't the vertical "wrong"? there are lines like
```
Tato^Itento^IPDFS1----------^IAtr^I+1^Ivyhláška^Ivyhláška^INNFS1-----A----^ISb^I+1^Ivyhláška^Ivyhláška^INNFS1-----A----^ISb$
```
seems there are 3 word/lemma/tag/afun (the `^I` are inserted by vim) on the line? maybe that's intended and handled in config?

**Comment by Ansa211** [comment]
The vertical is correct, the three sets belong to the word itself (attributes `word,lemma,tag,afun`), its parent (attributes `p_form,p_lemma,p_tag,p_afun`) and its effective parent (attributes `ep_form,ep_lemma,ep_tag,ep_afun`). Most of the time, the parent and effective parent sets have the same values.

**Comment by Ansa211** [comment]
I cannot reproduce this on UČNK.

**Comment by Ansa211** [comment]
Note:

[`[lc="vyhláška"|lemma="(?i)vyhláška"]`](http://lindat.mff.cuni.cz/services/kontext/first_form?default_attr=word&corpname=legaltext_cs_a&usesubcorp=&queryselector=cqlrow&cql=%5Blc%3D%22vyhl%C3%A1%C5%A1ka%22%7Clemma%3D%22%28%3Fi%29vyhl%C3%A1%C5%A1ka%22%5D) gives wrong results, but 
[`[lemma="(?i)vyhláška"|lc="vyhláška"]`](http://lindat.mff.cuni.cz/services/kontext/first_form?default_attr=word&corpname=legaltext_cs_a&usesubcorp=&queryselector=cqlrow&cql=%5Blemma%3D%22%28%3Fi%29vyhl%C3%A1%C5%A1ka%22%7Clc%3D%22vyhl%C3%A1%C5%A1ka%22%5D) does not - the order of the disjuncts matters!
[`[lc="vyhláška"|lemma="vyhláška"]`](http://lindat.mff.cuni.cz/services/kontext/first_form?default_attr=word&corpname=legaltext_cs_a&usesubcorp=&queryselector=cqlrow&cql=%5Blc%3D%22vyhl%C3%A1%C5%A1ka%22%7Clemma%3D%22vyhl%C3%A1%C5%A1ka%22%5D) is also OK, and even more surprisingly
[`[lc=".*hláška"|lemma="(?i).*hláška"]`](http://lindat.mff.cuni.cz/services/kontext/first_form?default_attr=word&corpname=legaltext_cs_a&usesubcorp=&queryselector=cqlrow&cql=%5Blc%3D%22.%2Ahl%C3%A1%C5%A1ka%22%7Clemma%3D%22%28%3Fi%29.%2Ahl%C3%A1%C5%A1ka%22%5D) gives the correct result, too (there are no other word forms/lemmas matching the regexp).

It's the same with basic search for `dům`: changing even just a single character in the corresponding query solves the problem, e.g. `[lc=".ům"|lemma="(?i)dům"]`

**Comment by Ansa211** [comment]
SOLUTION: The problem is that after recompiling a corpus (on kontext-dev) or updating compiled index files in some other way (on production), the cached concordances should be removed. When I did this, I first got a "missing cache file" message, but after running the same query for a second time, the result was correct.

The problem disappears after deleting the problematic corpus from these directories:
kontext-new: `/export/KONTEXT/kontext/cache-new/conc`
kontext-dev: `/opt/kontext-ansa/cache/vyhlasky_cs_a/`

On kontext-dev, I can solve the problem by adding a "remove the cache" block in the compilation pipeline.
However, **for avoiding a similar problem on production in the future, I would suggest that the internals of KonText should make sure that it only uses cache files that are newer than the index files used**. Is that doable and reasonable?
